specs/merkle-tree.md

```diff
@@ -0,0 +1,142 @@
+
+### 4. There must be exactly one `APPROVE` leaf or `CANCEL` leaf per chain
+Constructing a tree that contains `APPROVE` or `CANCEL` leaves that approve or cancel multiple separate deployments on a single chain would lead to an ambiguous situation for the relayer where it is unclear which deployment should be approved or canceled. So we impose the restriction that in a valid Merkle tree there must be exactly one `APPROVE` or `CANCEL` leaf per chain.
+
+If the `arbitraryChain` field is set to true, then we consider this approval leaf to apply to all chains. Therefore if `arbitraryChain` = true, there should be exactly one `APPROVE` leaf in the entire tree and no `CANCEL` leaves.
+
+### 5. Merkle tree leaves must be ordered in the tree by the leaf's `index` and `chainId` fields ascending [^2]
+Sphinx Merkle tree leaves use an explicit `index` field in the leaf rather than an implicit index determined by the leaves position in the tree. This is because the Sphinx Merkle tree is intended to be executed across multiple networks and not all leaves will be executed on all networks. We impose the restriction the tree leaves will be ordered by their `index` and `chainId` fields ascending.
+
+So for example, if we intended to run a single transaction on Goerli (5), OP Goerli (420), and Polygon Mumbai (80001). Then the leaves are are expected to be ordered like so:
+1. Approval Leaf on Goerli (Chain ID 5)
+2. Execute Leaf on Goerli
+3. Approval Leaf on OP Goerli (Chain ID 420)
+4. Execute Leaf on OP Goerli
+5. Approval Leaf on Polyon Mumbai (Chain ID 80001)
+6. Execute Leaf on Polygon Mumbai
+
+## Merkle Tree Generation Logic
+We provide utility for generating Sphinx Merkle trees. This utility is used by the official Sphinx Foundry plugin. We expect it to generate Merkle trees that adhere to the above invariants.
+
+#### `const makeSphinxMerkleTree = (deploymentData: DeploymentData): SphinxMerkleTree`
+
+#### Input
+Accepts a [DeploymentData](https://github.com/sphinx-labs/sphinx/blob/591ff8a33578ecf61f184f4afc83a551397009f9/packages/contracts/src/merkle-tree.ts#L45C23-L45C23) object where the keys are canonical chain ids and the values are deployment data objects which contain all of the necessary info to assemble a SphinxMerkleTree. The Sphinx Merkle tree generation function should be agnostic to the source of the transaction data so it can be used to assemble a tree based on transactions generated from any scripting framework (i.e Foundry, Hardhat Ignition, some arbitrary future framework).
+
+#### Output
+Outputs a [SphinxMerkleTree](https://github.com/sphinx-labs/sphinx/blob/591ff8a33578ecf61f184f4afc83a551397009f9/packages/contracts/src/merkle-tree.ts#L115) object that follows the above specification and invariants.
```

Comment:
> above specification and invariants

I think this should be a bit more precise. Unless there are specific parts of the spec that you're referring to, I think we should just say "invariants"